src/fiber/fiber.ml

```diff
@@ -604,51 +567,86 @@ end
 
 module Stream = struct
   module In = struct
-    type nonrec 'a t = unit -> 'a option t
+    (* Invariant: once [read] has returned [None], it always returns [None] *)
+    type nonrec 'a t =
+      { mutable read : unit -> 'a option t
+      ; mutable reading : bool
+      }
+
+    let create_unchecked read = { read; reading = false }
+
+    let create read =
+      let t = { read; reading = false } in
+      let read () =
+        let+ x = read () in
+        if Option.is_none x then t.read <- (fun () -> return None);
+        x
+      in
+      t.read <- read;
+      t
+
+    let lock t =
+      if t.reading then Code_error.raise "Fiber.Stream.In: already reading" [];
+      t.reading <- true
 
-    let create f = f
+    let unlock t = t.reading <- false
 
-    let read t = t ()
+    let read t =
```

Comment:
I don't think it matters for my uses, but why isn't it allowed to call multiple reads? Sure, it might not be correct for every implementation of a stream, but the user can easily implement it:
```
let in_ =
  let creading = ref false in
  create (fun () ->
    if !reading then
      Code_error.raise
   else
    let ( ) = reading := true in
     read stream)
```

Reply:
Because of the test added in the second commit. The behaviour is odd. We should either add a mutex in stream to sequentialise read/write, or fail as I did here. We can see if the failing behaviour is annoying in practice (especially for the write side).

Reply:
The behavior did end up being annoying in practice for writers. In LSP, I have a `Fiber.Stream.Out.t` that is backed by a thread writing to a channel. So the writes are already sequentialised automatically.

By the way, the current behavior is also easy enough to replicate by users:

```ocaml
Fiber.Stream.Out.create (
  let writing = ref false in
  fun e ->
    if !r then failwith "already writing";
    r := true;
   ..
)
```